### PR TITLE
Fixing the total loss function calculation

### PR DESCRIPTION
**Pull request by sreeshnair**
Pull Request Checklist
 - [x] All tests in the `tests` folder pass with a local build
 - [x] Pull request includes a description of why we are doing this
 - [x] Init files import new capabilities to appropriate level of package (if applicable)
 - [x] CHANGELOG has been updated
 - [x] Version in `_version.py` has been updated
 - [x] README has been updated (if applicable)
 - [x] `requirements.txt` and `requirements-dev.txt` have been recompiled if requirements in `setup.py` or `requirements-dev.in` changed.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,7 +2,10 @@
 All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project uses [Semantic Versioning](http://semver.org/).
-
+# [3.1.6] - 2021-10-22
+  ### Changed
+  - Optimized the total loss functions
```
Comment (gsganden) [inline CHANGELOG.md:7]:
```suggestion
  - Use a running average for total training loss as well as individual task training loss
```

File: octopod/learner.py [hunk 2]
```diff
@@ -71,14 +71,14 @@ def __init__(self,
...
-        best_model=False,
-        smooth_loss_alpha=0.2
+            self,
+            num_epochs,
+            scheduler,
+            step_scheduler_on_batch,
+            optimizer,
+            device='cuda:0',
+            best_model=False,
+            smooth_loss_alpha=0.2
```
Comment (gsganden) [inline octopod/learner.py:81]: Please restore the original 4-space indent.

File: octopod/learner.py [hunk 3]
```diff
@@ -255,7 +270,7 @@ def validate(self, device='cuda:0', pbar=None):
 
         with torch.no_grad():
             for step, batch in enumerate(
-                progress_bar(self.val_dataloader, parent=pbar, leave=(pbar is not None))
+                    progress_bar(self.val_dataloader, parent=pbar, leave=(pbar is not None))
```
Comment (gsganden) [inline octopod/learner.py:273]: Please restore the original 4-space indent.

**Review by gsganden — CHANGES_REQUESTED** [review]
These changes are looking nice!

I'm wondering if it's more useful to give each row equal weight (as this PR does) or to give each task equal weight. I'm leaning toward giving each task equal weight so that you're more likely to notice when there's a problem even if it's on a task with relatively few rows. @sugi-chan or @parsing-science do you have a preference?

**Comment by parsing-science** [comment]
@gsganden: I think giving each task equal weight makes sense. Some rows may be in multiple tasks so it feels like those might get overweighted if it's by row.

File: octopod/learner.py [hunk 4]
```diff
@@ -204,14 +203,30 @@ def fit(
...
             print(f'Epoch {best_model_epoch} best model saved with loss of {current_best_loss}')
 
-    def _update_smooth_training_loss_dict(self, task_type, current_loss, smooth_loss_alpha):
+    def _calculate_overall_loss(self):
+        overall_loss_total = sum(
+            self.smooth_training_item_cnt_dict[task] * self.smooth_training_loss_dict[task]
+            for task in self.smooth_training_loss_dict
+        )
+
+        return overall_loss_total / sum(self.smooth_training_item_cnt_dict.values())
```
Comment (gsganden) [inline octopod/learner.py:212]: As we discussed earlier, I think we can have this function just average `smooth_training_loss_dict` without weighting by rows, and then we will not need `smooth_training_item_cnt_dict`.

Comment (sreeshnair) [reply octopod/learner.py]: removed the new attribute
